torch/cuda/amp/autocast_mode.py

```diff
@@ -0,0 +1,189 @@
+import torch
+import functools
+import warnings
+import numpy as np
+from torch._six import container_abcs, string_classes
+
+
+class autocast(object):
+    r"""
+    Instances of :class:`autocast` serve as context managers or decorators that
+    allow regions of your script to run in mixed precision.
+
+    In autocast-enabled regions, the backend automatically chooses the precision
+    for GPU operations to improve performance while maintaining accuracy.
+
+    When entering an autocast-enabled region, Tensors may be any type.  It is not necessary or
+    recommended to call ``.half()`` on your model(s) or data to use autocasting.
+
+    :class:`autocast` should wrap the forward pass(es) of your network::
```

Comment:
Should wrap the forward passes and the loss computation? I don't know if loss computation is always considered part of forward. In your example it looks like a separate part of the program.